pkg/store/bucket.go

```diff
@@ -856,8 +856,42 @@ func chunksSize(chks []storepb.AggrChunk) (size int) {
 }
 
 // LabelNames implements the storepb.StoreServer interface.
-func (s *BucketStore) LabelNames(context.Context, *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *BucketStore) LabelNames(ctx context.Context, r *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	s.mtx.RLock()
+
+	var mtx sync.Mutex
+	var sets [][]string
+
+	for _, b := range s.blocks {
+		indexr := b.indexReader(gctx)
+		g.Go(func() error {
+			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")
+
+			res, err := indexr.LabelNames(gctx)
+			sort.Strings(res)
+			mtx.Lock()
+			sets = append(sets, res)
+			mtx.Unlock()
+			return err
+		})
+	}
+
+	s.mtx.RUnlock()
+
+	if err := g.Wait(); err != nil {
+		if !r.PartialResponseDisabled {
```

Comment:
Please kill this, just aborting is fine. We are not in Querier.

Essentially If we have error here, means something serious on store gateway indexes and our version is not TSDB compatibile.